oslo_vmware/image_transfer.py

```diff
@@ -346,3 +360,58 @@ def upload_image(context, timeout_secs, image_service, image_id, owner_id,
         updater.stop()
         read_handle.close()
     LOG.debug("Uploaded image: %s.", image_id)
+
+
+def image_pull_from_swift(swift_url, **kwargs):
+    (lease, lease_info) = \
+        rw_handles.VmdkWriteHandle._create_import_vapp_lease(
+            kwargs.get('session'), kwargs.get('resource_pool'),
+            kwargs.get('vm_import_spec'), kwargs.get('vm_folder'))
+
+    return image_pull_from_url(
+        kwargs.get('session'), lease, lease_info, swift_url.url,
+        size=kwargs.get('image_size'),
+        ssl_thumbprint=swift_url.ssl_thumbprint(),
+        headers={'X-Auth-Token': swift_url.auth_token})
+
+
+def image_pull_from_url(session, lease_ref, lease_info, url,
+                        size=None, ssl_thumbprint=None, headers=None):
+    """Converts a lease to pull from URLs and waits for the import.
+    Using HttpNfcLeasePullFromUrls_Task
+    """
+    if not lease_info.deviceUrl:
+        raise Exception("Invalid HttpNfc lease. "
+                        "No DeviceURLs found to import to.")
+
+    import_key = lease_info.deviceUrl[0].importKey
+
+    client_factory = session.vim.client.factory
+    file_spec = client_factory.create("ns0:HttpNfcLeaseSourceFile")
+    file_spec.targetDeviceId = import_key
+    file_spec.url = url
+    file_spec.create = True
+    file_spec.sslThumbprint = ssl_thumbprint
+    if size:
+        file_spec.size = size
+    if headers:
+        file_spec.httpHeaders = vim_util.dict_to_kv(client_factory,
+                                                    headers)
+    pull_task = session.invoke_api(session.vim,
+                                   "HttpNfcLeasePullFromUrls_Task",
+                                   lease_ref,
+                                   files=[file_spec])
+
+    LOG.debug("Started pulling the image from URL %(url)s",
+              {'url': url})
+
```

Comment:
If the lease is converted to a pull, does it not require updating anymore?

Reply:
Operations taking 20+ minutes went OK.
I would expect VMware to take care of the lease, since it is the one using it, we're not using the lease anymore.
I can try however updating the lease, to see what happens and if it still accepted.

Reply:
I believe that to be true too. Just wanted to make sure. Do you know the timeout value? If it is lower than 20min then I am fine with that.

Reply:
Which timeout do you mean? There is no timeout that I am aware of here. Basically `wait_for_task` will return only when VMware finished downloading the image internally.

Reply:
I mean the timeout the vcenter has before it expires a lease due to a lack of update via `HttpNfcLeaseProgress`.

Reply:
This is not clearly documented anywhere. From the VSphere API documentation:
> A lease will time out automatically after a while. If the client wishes to continue using it, for example if it is not done accessing the disks, this method must be called periodically.
... but we don't want to access the disk through the lease in this case.

The upstream nova driver updates the lease every 60 seconds.

Reply:
Okay, then let's hope pulling shifts the responsibility to the vsphere service.